tensorflow/contrib/seq2seq/python/ops/layers.py

```diff
@@ -20,15 +20,185 @@
 from __future__ import division
 from __future__ import print_function
 
+from tensorflow.python.ops import tensor_array_ops
 from tensorflow.python.ops import array_ops
+from tensorflow.python.ops import variable_scope as vs
+from tensorflow.python.ops import math_ops
+from tensorflow.python.ops import control_flow_ops
+from tensorflow.python.ops import rnn
 
+from tensorflow.contrib.layers import fully_connected
+
+from functools import partial
 
 __all__ = ["rnn_decoder",
            "rnn_decoder_attention"]
 
+"""Used to project encoder state in `rnn_decoder`"""
+encoder_projection = partial(fully_connected, activation_fn=math_ops.tanh)
 
-def rnn_decoder(*args, **kwargs):
-  pass
+
+def rnn_decoder(cell, decoder_inputs, initial_state,
+                sequence_length, decoder_fn,
+                encoder_projection=encoder_projection,
+                parallel_iterations=None, swap_memory=False,
+                time_major=False, scope=None):
+  """RNN decoder for a sequence-to-sequence model specified by RNNCell 'cell'.
+
+  The 'rnn_decoder' is similar to the 'tf.python.ops.rnn.dynamic_rnn'. As the
+  decoder does not make any assumptions of sequence length of the input or how
+  many steps it can decode, since 'rnn_decoder' uses dynamic unrolling. This
+  allows `decoder_inputs` to have [None] in the sequence length of the decoder
+  inputs.
+
+  The parameter decoder_inputs is nessesary for both training and evaluation.
+  During training it is feed at every timestep. During evaluation it is only
+  feed at time==0, as the decoder needs the `start-of-sequence` symbol, known
+  from Sutskever et al., 2014 https://arxiv.org/abs/1409.3215, at the
+  beginning of decoding.
+
+  The parameter sequence length is nessesary as it determines how many
+  timesteps to decode for each sample. TODO: Could make it optional for
```

Comment:
While you're decoding, can't you choose to end decoding based on what you just decoded?  i.e., if you decoded an end-of-sentence token, shouldn't that mark being done?  In fact, it seems that `sequence_length` should only be necessary at training, never at eval.